File: camera/siyi_gimbal.py
import socket
import struct
import time

class SiyiGimbal:
    """
    SIYI A8 Mini 짐벌 제어 (C 코드 A8miniControl.c 호환 버전)
    
    C 코드 분석 결과:
    - CMD_ID: 0x07 (GIMBAL_SPEED)
    - Payload: Yaw(1byte) + Pitch(1byte) = Total 2 bytes
    - Data Range: -100 ~ 100 (int8)
    """
    def __init__(self, ip='192.168.144.25', port=37260, debug=False):
        self.ip = ip
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.debug = debug
        self.seq = 0
        
        # 제어 파라미터
        self.kp_x = 0.5  # Yaw 반응 속도
        self.kp_y = 0.5  # Pitch 반응 속도
        self.deadzone = 0.05

    def _crc16_cal(self, data, crc=0):
        """SIYI SDK CRC16 계산 (C 코드와 동일 로직)"""
        for byte in data:
            crc ^= byte
            for _ in range(8):
                if crc & 1:
                    crc = (crc >> 1) ^ 0xA001
                else:
                    crc = (crc >> 1)
        return crc

    def send_packet(self, cmd_id, data_bytes):
        """패킷 생성 및 전송"""
        STX = b'\x55\x66'
        CTRL = b'\x01' # C코드: 0x01 (Need ACK / ACK Pack)
        
        # 데이터 길이 (Data Length): 2 bytes, Little Endian
        # C코드에서는 0x02, 0x00 (2 bytes)
        data_len = len(data_bytes)
        DATA_LEN = struct.pack('<H', data_len)
        
        # 시퀀스 (SEQ): C코드에서는 0x0000으로 고정해서 보냄
        SEQ = struct.pack('<H', 0) 
        
        CMD_ID = struct.pack('<B', cmd_id)
        
        # CRC 계산을 위한 앞부분 (STX ~ DATA)
        packet_without_crc = STX + CTRL + DATA_LEN + SEQ + CMD_ID + data_bytes
        
        # CRC 계산
        crc_val = self._crc16_cal(packet_without_crc, 0)
        CRC = struct.pack('<H', crc_val)
        
        final_packet = packet_without_crc + CRC
        
        try:
            self.sock.sendto(final_packet, (self.ip, self.port))
            if self.debug:
                # C 코드 디버그 출력과 비교하기 쉽도록 HEX 출력
                hex_str = " ".join([f"{b:02x}" for b in final_packet])
                print(f"[TX] {hex_str}")
        except Exception as e:
            print(f"[SIYI] Send Error: {e}")

    def send_speed(self, yaw_speed, pitch_speed):
        """
        짐벌 속도 제어
        CMD_ID: 0x07 (GIMBAL_SPEED)
        Payload: [Yaw(int8)][Pitch(int8)]
        """
        # 범위 제한 (-100 ~ 100)
        yaw_speed = int(max(-100, min(100, yaw_speed)))
        pitch_speed = int(max(-100, min(100, pitch_speed)))
        
        # [핵심 수정] C코드와 동일하게 'b' (signed char, 1byte) 사용
        # struct.pack('<bb', ...) -> 2바이트 페이로드 생성
        data = struct.pack('<bb', yaw_speed, pitch_speed)
        
        # 0x07 명령어로 전송
        self.send_packet(0x07, data)

    def track_object(self, bbox, w, h):
        """객체 추적 로직 (자동 호출됨)"""
        if not bbox:
            self.send_speed(0, 0)
            return

        # Hailo BBox (0.0 ~ 1.0 정규화 좌표)
        ymin, xmin, ymax, xmax = bbox
        cx = (xmin + xmax) / 2.0
        cy = (ymin + ymax) / 2.0
        
        # 중앙(0.5)에서의 오차 계산
        err_x = cx - 0.5
        err_y = cy - 0.5
        
        # 데드존(Deadzone) 체크
        if abs(err_x) < self.deadzone: err_x = 0
        if abs(err_y) < self.deadzone: err_y = 0
        
        # 속도 계산 (P 제어)
        # 100을 곱해서 -100~100 범위로 매핑
        # 방향이 반대면 부호(-)를 조정하세요
        yaw_cmd = err_x * 100 * self.kp_x
        pitch_cmd = -err_y * 100 * self.kp_y 
        
        self.send_speed(yaw_cmd, pitch_cmd)

# 단독 실행 테스트
if __name__ == "__main__":
    g = SiyiGimbal(debug=True)
    print("Testing Gimbal Move (Right 15)...")
    
    # 0.1초 간격으로 10번 전송 (UDP 특성상 반복 전송 권장)
    for _ in range(10):
        g.send_speed(15, 0) 
        time.sleep(0.1)
        
    print("Stopping...")
    g.send_speed(0, 0)

File: camera/siyi_gimbal.c
// siyi_gimbal.c
// SIYI A8 Mini gimbal control (UDP)
// Usage:
//   1) Auto center:
//      ./siyi_gimbal
//   2) Tracking with bbox (normalized coords):
//      ./siyi_gimbal <ymin> <xmin> <ymax> <xmax>

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <math.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>

// Gimbal network settings
#define GIMBAL_IP   "192.168.144.25"
#define GIMBAL_PORT 37260

// Control parameters
#define DEADZONE    0.05f
#define KP_X        0.5f
#define KP_Y        0.5f

// CRC16-CCITT (poly = 0x1021, init = 0x0000, MSB-first)
static uint16_t crc16_ccitt(const uint8_t *data, size_t len, uint16_t crc_init)
{
    uint16_t crc = crc_init & 0xFFFF;
    const uint16_t poly = 0x1021;

    for (size_t i = 0; i < len; ++i) {
        crc ^= (uint16_t)(data[i] << 8);
        for (int bit = 0; bit < 8; ++bit) {
            if (crc & 0x8000) {
                crc = (uint16_t)((crc << 1) ^ poly);
            } else {
                crc <<= 1;
            }
        }
    }
    return crc & 0xFFFF;
}

static void print_hex(const uint8_t *buf, size_t len)
{
    for (size_t i = 0; i < len; ++i) {
        printf("%02X ", buf[i]);
    }
    printf("\n");
}

// Generic packet builder
static void send_siyi_packet(int sockfd,
                             struct sockaddr_in *addr,
                             uint8_t cmd_id,
                             const uint8_t *data,
                             uint16_t data_len)
{
    uint8_t packet[32];
    size_t idx = 0;

    // Header: STX
    packet[idx++] = 0x55;
    packet[idx++] = 0x66;

    // CTRL
    packet[idx++] = 0x01;

    // DATA_LEN (little endian)
    packet[idx++] = (uint8_t)(data_len & 0xFF);
    packet[idx++] = (uint8_t)((data_len >> 8) & 0xFF);

    // SEQ (0)
    packet[idx++] = 0x00;
    packet[idx++] = 0x00;

    // CMD_ID
    packet[idx++] = cmd_id;

    // DATA
    for (uint16_t i = 0; i < data_len; ++i) {
        packet[idx++] = data[i];
    }

    // CRC16 over all bytes so far
    uint16_t crc = crc16_ccitt(packet, idx, 0x0000);
    packet[idx++] = (uint8_t)(crc & 0xFF);        // low
    packet[idx++] = (uint8_t)((crc >> 8) & 0xFF); // high

    // Send via UDP
    sendto(sockfd, packet, idx, 0, (struct sockaddr *)addr, sizeof(*addr));

    printf("[SIYI TX] cmd=0x%02X len=%zu | ", cmd_id, idx);
    print_hex(packet, idx);
}

// Auto center command (CMD 0x08, DATA 0x01)
static void send_auto_center(int sockfd, struct sockaddr_in *addr)
{
    uint8_t data[1] = {0x01};
    send_siyi_packet(sockfd, addr, 0x08, data, 1);
}

// Speed command (CMD 0x07, DATA = yaw(int8), pitch(int8))
static void send_speed(int sockfd,
                       struct sockaddr_in *addr,
                       int yaw_speed,
                       int pitch_speed)
{
    // Clamp range to [-100, 100]
    if (yaw_speed > 100) yaw_speed = 100;
    if (yaw_speed < -100) yaw_speed = -100;
    if (pitch_speed > 100) pitch_speed = 100;
    if (pitch_speed < -100) pitch_speed = -100;

    int8_t yaw = (int8_t)yaw_speed;
    int8_t pitch = (int8_t)pitch_speed;

    uint8_t data[2];
    data[0] = (uint8_t)yaw;
    data[1] = (uint8_t)pitch;

    printf("[SIYI] Speed command yaw=%d pitch=%d\n", yaw_speed, pitch_speed);
    send_siyi_packet(sockfd, addr, 0x07, data, 2);
}

int main(int argc, char *argv[])
{
    int sockfd;
    struct sockaddr_in servaddr;

    // Create UDP socket
    if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        perror("socket creation failed");
        return -1;
    }

    memset(&servaddr, 0, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_port   = htons(GIMBAL_PORT);
    servaddr.sin_addr.s_addr = inet_addr(GIMBAL_IP);

    // Case 1: no arguments -> auto center
    if (argc == 1) {
        printf("[INFO] No bbox provided -> auto center.\n");
        send_auto_center(sockfd, &servaddr);
        close(sockfd);
        return 0;
    }

    // Case 2: bbox mode (ymin, xmin, ymax, xmax)
    if (argc != 5) {
        fprintf(stderr,
                "Usage:\n"
                "  %s                 # auto center\n"
                "  %s <ymin> <xmin> <ymax> <xmax>  # tracking (normalized)\n",
                argv[0], argv[0]);
        close(sockfd);
        return -1;
    }

    float ymin = strtof(argv[1], NULL);
    float xmin = strtof(argv[2], NULL);
    float ymax = strtof(argv[3], NULL);
    float xmax = strtof(argv[4], NULL);

    // Compute bbox center in normalized coordinates
    float cx = (xmin + xmax) * 0.5f;
    float cy = (ymin + ymax) * 0.5f;

    // Error from image center (0.5, 0.5)
    float err_x = cx - 0.5f;
    float err_y = cy - 0.5f;

    // Apply deadzone
    if (fabsf(err_x) < DEADZONE) err_x = 0.0f;
    if (fabsf(err_y) < DEADZONE) err_y = 0.0f;

    int yaw_cmd   = (int)(err_x * 100.0f * KP_X);
    int pitch_cmd = (int)(-err_y * 100.0f * KP_Y);  // invert pitch if needed

    printf("[INFO] bbox=(%.3f, %.3f, %.3f, %.3f), center=(%.3f, %.3f), "
           "err=(%.3f, %.3f), cmd=(%d, %d)\n",
           ymin, xmin, ymax, xmax, cx, cy, err_x, err_y, yaw_cmd, pitch_cmd);

    // Send speed command
    send_speed(sockfd, &servaddr, yaw_cmd, pitch_cmd);

    close(sockfd);
    return 0;
}
